gdnative-core/src/user_data.rs

```diff
@@ -0,0 +1,386 @@
+//! no UB can occur because we freed `owner` or put another script on it.
+//! 
+//! ## Which wrapper to use?
+//! 
+//! ### Use a `MutexData<T>` when:
+//! 
+//! - You don't want to handle locks explicitly.
+//! - Your `NativeClass` type is only `Send`, but not `Sync`.
+//! 
+//! ### Use a `RwLockData<T>` when:
+//! 
+//! - You don't want to handle locks explicitly.
+//! - Some of your exported methods take `&self`, and you don't need them to be exclusive.
+//! - Your `NativeClass` type is `Send + Sync`.
+//! 
+//! ### Use a `ArcData<T>` when:
+//! 
+//! - You want safety for your methods, but can't tolerate lock overhead on each method call.
+//! - You want fine grained lock control for parallelism.
+//! - All your exported methods take `&self`.
+//! - Your `NativeClass` type is `Send + Sync`.
+
+use std::mem;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use std::marker::PhantomData;
+use parking_lot::{Mutex, RwLock};
+
+use crate::NativeClass;
+
+/// Trait for customizable user-data wrappers.
+/// 
+/// See module-level documentation for detailed explanation on user-data.
+pub unsafe trait UserData: Sized + Clone {
+    type Target: NativeClass;
+
+    /// Creates a new owned wrapper from a `NativeClass` instance.
+    fn new(val: Self::Target) -> Self;
+
+    /// Takes a native instance and returns an opaque pointer that can be used to recover it.
+    /// 
+    /// This gives "ownership" to the engine.
+    unsafe fn into_user_data(self) -> *const libc::c_void;
+
+    /// Takes an opaque pointer produced by `into_user_data` and "consumes" it to produce the
+    /// original instance, keeping the reference count.
+    /// 
+    /// This should be used when "ownership" is taken from the engine, i.e. destructors.
+    /// Use elsewhere can lead to premature drops of the instance contained inside.
+    unsafe fn consume_user_data_unchecked(ptr: *const libc::c_void) -> Self;
+
+    /// Takes an opaque pointer produced by `into_user_data` and "clones" it to produce the
+    /// original instance, increasing the reference count.
+    /// 
+    /// This should be used when user data is "borrowed" from the engine.
+    unsafe fn clone_from_user_data_unchecked(ptr: *const libc::c_void) -> Self;
+}
+
+/// Trait for wrappers that can be mapped immutably.
+pub trait Map: UserData {
+    type Err: Debug;
+
+    /// Maps a `&T` to `U`. Called for methods that take `&self`.
+    fn map<F, U>(&self, op: F) -> Result<U, Self::Err>
+    where
+        F: FnOnce(&Self::Target) -> U;
+}
+
+/// Trait for wrappers that can be mapped mutably.
+pub trait MapMut: UserData {
+    type Err: Debug;
+
+    /// Maps a `&mut T` to `U`. Called for methods that take `&mut self`.
+    fn map_mut<F, U>(&self, op: F) -> Result<U, Self::Err>
+    where
+        F: FnOnce(&mut Self::Target) -> U;
+}
+
+/// Error type indicating that an operation can't fail.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub enum Infallible { }
```

Comment:
Could we use the `core::conv::Infallible` for this?

Reply:
I don't think so. `conv::Infallible` isn't a general purpose `never` type, but only for type conversions. Practically, all empty enums are the same. However, the `map` functions aren't supposed to actually do any type conversions (it should be done at the constructor), so semantically that would be a tiny little bit nonsense.

Ideally, we'd like to use `!`, but that's not stable yet, so I think it's currently best to create a separate type.

Reply:
Yes, fair point!